http4s/src/main/scala/io/isomarcte/errors4s/http4s/ResponseError.scala

```diff
@@ -0,0 +1,144 @@
+    ErrorBody
+      .fromResponse_(maxBodySize)(response)
+      .map(errorBody =>
+        ResponseError(response.status, response.headers, errorBody, maxBodySize, requestUri, requestMethod)
+      )
+
+  /** Create a [[ResponseError]] which will read, at most, the default number of
+    * bytes from the error response, given by
+    * [[ErrorBody#defaultMaxBodySizeBytes]].
+    */
+  def fromResponse[F[_]](requestUri: Option[Uri], requestMethod: Option[Method])(
+    response: Response[F]
+  )(implicit C: Stream.Compiler[F, F], F: FlatMap[F]): F[Throwable] =
+    fromResponse_(ErrorBody.defaultMaxBodySizeBytes, requestUri, requestMethod)(response)
+
+  /** Algebraic Data Type for error response bodies.
+    *
+    * Members include, [[ErrorBody#FullErrorBody]] for when the entire error
+    * body is smaller than the given max bytes limit,
+    * [[ErrorBody#PartialErrorBody]] for when the error body is larger than
+    * the given max bytes limit, and [[ErrorBody#EmptyErrorBody]] for when
+    * there is no body or the body's size is 0 bytes.
+    *
+    * @note There are a few edge cases here to be aware of. If the error body
+    *       is ''exactly'' the size of the `maxBodySize` value, then it will
+    *       be a [[ErrorBody#PartialErrorBody]]. This is because it can't be
+    *       known if there is another byte without reading another byte, which
+    *       would break the contract of this type. If the `maxBodySize` is
+    *       `0L` then the result will be a
+    *       [[ErrorBody#PartialErrorBody]]. This is because we were
+    *       ''technically'' able to read the specified max number of
+    *       bytes. That being said, there is no reason to have `maxBodySize`
+    *       be <= 0.
+    */
+  sealed trait ErrorBody extends Product with Serializable {
+    import ErrorBody._
+
+    final lazy val asText: Either[Throwable, Option[String]] =
+      this match {
+        case FullErrorBody(value) =>
+          value.decodeUtf8.map(_.pure[Option]).widen
+        case PartialErrorBody(value) =>
+          value.decodeUtf8.map(_.pure[Option]).widen
+        case EmptyErrorBody =>
+          Right(None)
+      }
+  }
+
+  object ErrorBody {
+
+    /** Default max body size in bytes, 1MiB. */
+    lazy val defaultMaxBodySizeBytes: Long = 1024L * 1024L // 1MiB
+
+    final case class FullErrorBody(value: ByteVector)    extends ErrorBody
+    final case class PartialErrorBody(value: ByteVector) extends ErrorBody
+    case object EmptyErrorBody                           extends ErrorBody
+
+    def fromResponse_[F[_]](
+      maxBodySize: Long
+    )(response: Response[F])(implicit C: Stream.Compiler[F, F], F: FlatMap[F]): F[ErrorBody] =
+      response
+        .body
+        .take(maxBodySize)
+        .compile
+        .to(ByteVector)
+        .map { (body: ByteVector) =>
+          val size: Long = body.size
+          if (size === maxBodySize || maxBodySize <= 0L) {
+            PartialErrorBody(body)
```

Comment:
This is a little surprising, but I _think_ I get it. If `maxBodySize <= 0L`, then the body should also have 0 bytes, which makes me think that it should be `EmptyErrorBody`, but at the same time, there _may_ have been more bytes that got cut off because of `maxBodySize`, so it _might_ be `PartialErrorBody`.  

There is one thing that might be confusing in that there's no check to see if any bytes _would_ have come through, so it actually might have been an empty error body after all. I'm not sure if that matters right now, but I can see it being an issue at some point.

Reply:
Yeah, I put in a comment about this. We can't validate if there were more bytes or not without trying to read them, which is against the contract.